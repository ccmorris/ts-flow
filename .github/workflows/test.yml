# Run unit tests with Bun on every push

name: CI Tests

on:
  push:
    branches:
      - '*'
  pull_request:
    branches:
      - main

jobs:
  test:
    runs-on: ubuntu-latest

    steps:
      - name: Checkout repository
        uses: actions/checkout@v3

      - name: Set up Bun
        uses: oven-sh/setup-bun@v1
        with:
          bun-version: latest # You can specify a version like '1.0.0' if needed

      - name: Install dependencies
        run: bun install

      - name: Run tests
        run: bun test
