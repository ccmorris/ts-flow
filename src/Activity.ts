import type {
  TaskDefinitions,
  ActivityFunction,
  EndPointer,
  CatchInput,
  ActivityContext,
} from './types'
import type { Choice } from './Choice'

/**
 * An activity is a single unit of work in a workflow.
 * Define an async function, and chain activities together to create a workflow.
 */
export class Activity<I, O, C extends ActivityContext> {
  name: string
  fn: ActivityFunction<I, O, C>
  next?: Activity<O, unknown, C> | Choice<O, unknown, C> | EndPointer
  catchConfig?: {
    [key: string]: {
      then:
        | Activity<CatchInput<any>, unknown, C>
        | Choice<CatchInput<any>, unknown, C>
        | EndPointer
    }
  }

  public constructor(name: string, fn: ActivityFunction<I, O, C>) {
    this.name = name
    this.fn = fn
  }

  /**
   * Chain another activity or choice to run after this one, when successful.
   */
  public then<T extends Activity<O, any, C> | Choice<O, any, C>>(next: T): T {
    this.next = next
    return next
  }

  /**
   * Chain a catch activity or choice to run after this one, when an error occurs.
   *
   * @example Catch an ErrorName and run an activity:
   * activity.catch('ErrorName', new Activity('activity', async () => {}))
   *
   * @example Catch an ErrorName and end the workflow as a success:
   * activity.catch('ErrorName', null)
   */
  public catch<CatchKey extends string>(
    error: CatchKey,
    next:
      | Activity<CatchInput<CatchKey>, any, C>
      | Choice<CatchInput<CatchKey>, any, C>
      | EndPointer
  ) {
    this.catchConfig = { ...this.catchConfig, [error]: { then: next } }
    return this
  }

  /**
   * Convert the activity and any chained activities to a list of task definitions.
   */
  public toTaskDefinitions(): TaskDefinitions {
    const tasks = [
      {
        type: 'activity' as const,
        name: this.name,
        fn: this.fn,
        then: this.next?.name ?? null,
        catch: this.catchConfig
          ? Object.fromEntries(
              Object.entries(this.catchConfig).map(([error, catchTask]) => {
                return [error, { then: catchTask.then?.name ?? null }]
              })
            )
          : undefined,
      },
      ...(this.next ? this.next.toTaskDefinitions() : []),
      ...(this.catchConfig
        ? Object.values(this.catchConfig)
            .map((catchTask) => catchTask.then?.toTaskDefinitions())
            .filter((task) => !!task)
            .flat()
        : []),
    ]
    return removeDuplicateTasks(tasks as TaskDefinitions)
  }
}

// removes duplicates from an array using the name property
const removeDuplicateTasks = (tasks: TaskDefinitions) => {
  const seen = new Set()
  return tasks.filter((task) => {
    if (seen.has(task.name)) return false
    seen.add(task.name)
    return true
  })
}
