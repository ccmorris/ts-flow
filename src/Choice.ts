import type {
  TaskDefinitions,
  ActivityFunction,
  EndPointer,
  WorkflowResult,
} from './types'
import { run } from './run'
import { toDiagramPngUrl, toMermaid } from './diagrams'
import type { Activity } from './Activity'

export class Choice<I, O> {
  name: string
  fn: ActivityFunction<I, O>
  choices: {
    [key: string]: Activity<I, any> | Choice<I, any> | EndPointer
  }

  public constructor(name: string, fn: ActivityFunction<I, O>) {
    this.name = name
    this.fn = fn
    this.choices = {}
  }

  public choice(
    name: string,
    next: Activity<I, any> | Choice<I, any> | EndPointer
  ) {
    this.choices = { ...this.choices, [name]: next }
    return this
  }

  public toTaskDefinitions(): TaskDefinitions {
    const tasks = [
      {
        type: 'choice' as const,
        name: this.name,
        fn: this.fn,
        choices: this.choices
          ? Object.fromEntries(
              Object.entries(this.choices).map(([key, choiceTask]) => {
                return [key, choiceTask?.name ?? null]
              })
            )
          : {},
      },
      ...(this.choices
        ? Object.values(this.choices)
            .map((choice) => choice?.toTaskDefinitions())
            .filter((task) => !!task)
            .flat()
        : []),
    ]
    return removeDuplicateTasks(tasks as TaskDefinitions)
  }

  public async run(initialInput: I): Promise<WorkflowResult> {
    return run(this.toTaskDefinitions(), initialInput)
  }

  public toMermaid(): string {
    return toMermaid(this.toTaskDefinitions())
  }
  public toPngUrl(workflowResult?: WorkflowResult): string {
    return toDiagramPngUrl(this.toTaskDefinitions(), workflowResult)
  }
}

// removes duplicates from an array using the name property
const removeDuplicateTasks = (tasks: TaskDefinitions) => {
  const seen = new Set()
  return tasks.filter((task) => {
    if (seen.has(task.name)) return false
    seen.add(task.name)
    return true
  })
}
