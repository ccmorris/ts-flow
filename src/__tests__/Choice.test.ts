import { describe, expect, mock, test } from 'bun:test'

import { Choice } from '../Choice'
import { Activity } from '../Activity'

describe('Choice class', () => {
  test('should create a choice', () => {
    const choice = new Choice('id', async () => {})

    expect(choice.name).toBe('id')
    expect(choice.fn).toBeInstanceOf(Function)
  })

  test('should chain choices', () => {
    const activity1 = new Activity('activity1', async (_: 'in') => {})
    const activity2 = new Activity('activity2', async (_: 'in') => {})
    const choice = new Choice('id', async (_: 'in') => {})
      .choice('decision1', activity1)
      .choice('decision2', activity2)

    expect(choice.name).toBe('id')
    expect(choice.fn).toBeInstanceOf(Function)
    expect(choice.choices).toEqual({
      decision1: activity1,
      decision2: activity2,
    })
  })

  test('should convert to task definitions', () => {
    const activity1 = new Activity('activity1', async () => {})
    const activity2 = new Activity('activity2', async () => {})
    const choice = new Choice('id', async () => {})
      .choice('decision1', activity1)
      .choice('decision2', activity2)

    const taskDefinitions = choice.toTaskDefinitions()

    expect(taskDefinitions).toMatchSnapshot()
  })

  test('should run the choices based on the output of the choice function', async () => {
    const activity1 = new Activity('activity1', mock().mockResolvedValue(null))
    const activity2 = new Activity('activity2', mock().mockResolvedValue(null))
    const choice = new Choice('id', mock().mockResolvedValue('decision1'))
      .choice('decision1', activity1)
      .choice('decision2', activity2)

    await choice.run('initial input')

    expect(choice.fn).toHaveBeenCalledTimes(1)
    expect(choice.fn).toHaveBeenCalledWith('initial input', expect.any(Object))
    expect(activity1.fn).toHaveBeenCalled()
    expect(activity1.fn).toHaveBeenCalledWith(
      'initial input',
      expect.any(Object)
    )
    expect(activity2.fn).not.toHaveBeenCalled()
  })
})
