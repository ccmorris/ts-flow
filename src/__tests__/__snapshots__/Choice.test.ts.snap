// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Choice class should convert to task definitions 1`] = `
[
  {
    "choices": {
      "decision1": "activity1",
      "decision2": "activity2",
    },
    "fn": [Function: AsyncFunction],
    "name": "id",
    "start": true,
    "type": "choice",
  },
  {
    "catch": undefined,
    "fn": [Function: AsyncFunction],
    "name": "activity1",
    "start": true,
    "then": null,
    "type": "activity",
  },
  {
    "catch": undefined,
    "fn": [Function: AsyncFunction],
    "name": "activity2",
    "start": true,
    "then": null,
    "type": "activity",
  },
]
`;
